src/main/java/com/philips/research/bombase/core/nuget/domain/NugetClient.java

```diff
@@ -0,0 +1,105 @@
+import org.springframework.stereotype.Component;
+import retrofit2.Call;
+import retrofit2.Retrofit;
+import retrofit2.converter.jackson.JacksonConverterFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Locale;
+import java.util.Optional;
+
+@Component
+public class NugetClient {
+    // TODO: Can these mappers be cleaned up a bit more?
+    private static final ObjectMapper MAPPER = new ObjectMapper()
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.NON_PRIVATE)
+            .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
+    private static final ObjectMapper XMLMAPPER = new XmlMapper()
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.NON_PRIVATE);
+    private final NugetAPI restJson;
+    private final NugetAPI restXml;
+    private final URI baseURI;
+
+    NugetClient() {
+        this(URI.create("https://api.nuget.org/v3/"));
+    }
+
+    NugetClient(URI uri) {
+        final var retrofitJson = new Retrofit.Builder()
+                .baseUrl(uri.toASCIIString())
+                .addConverterFactory(JacksonConverterFactory.create(MAPPER))
+                .build();
+        final var retrofitXML = new Retrofit.Builder()
+                .baseUrl(uri.toASCIIString())
+                .addConverterFactory(JacksonConverterFactory.create(XMLMAPPER))
+                .build();
+        restJson = retrofitJson.create(NugetAPI.class);
+        restXml = retrofitXML.create(NugetAPI.class);
+        baseURI = uri;
+    }
+
+    Optional<PackageMetadata> getPackageMetadata(PackageURL purl) {
+        final var catalogResponse =
+                query(restJson.getCatalogEntry(purl.getName().toLowerCase(
+                        Locale.ROOT), purl.getVersion()));
+        if (catalogResponse.isEmpty() || catalogResponse.get().catalogEntry == null) {
+            return Optional.empty();
+        } else {
+            final var catalogEntryUrl = Optional.ofNullable(catalogResponse.get()
+                    .catalogEntry.split(baseURI.toASCIIString())[1]);
+            final var nugetSpecResponse =
+                    query(restXml.getNugetSpec(
+                    String.format("flatcontainer/%s/%s/%s.nuspec",
+                            purl.getName().toLowerCase(Locale.ROOT),
+                            purl.getVersion(),
+                            purl.getName().toLowerCase(Locale.ROOT))));
+            if (catalogEntryUrl.isPresent()) {
+                return query(restJson.getDefinition(catalogEntryUrl.get()))
+                        .map(result -> { result.downloadLocation =
+                                catalogResponse.get().packageContent;
+                    result.sourceUrl = nugetSpecResponse.get().metadata.repository.url;
```

Comment:
What happens in the (possibly unlikely) case when the `nugetSpecResponse` is empty? (I predict it will blow up.)

Perhaps more safe to use the actual state of the `Optional` value.